wal: TestFailoverWriter is flaky on windows

```
2024-02-12T20:39:05.8678369Z         	Error Trace:	D:/a/pebble/pebble/wal/failover_writer_test.go:227
2024-02-12T20:39:05.8678908Z         	            				D:/a/pebble/pebble/wal/failover_writer_test.go:350
2024-02-12T20:39:05.8680212Z         	            				C:/Users/runneradmin/go/pkg/mod/github.com/cockroachdb/datadriven@v1.0.3-0.20230413201302-be42291fc80f/datadriven.go:117
2024-02-12T20:39:05.8681482Z         	            				C:/Users/runneradmin/go/pkg/mod/github.com/cockroachdb/datadriven@v1.0.3-0.20230413201302-be42291fc80f/datadriven.go:336
2024-02-12T20:39:05.8682748Z         	            				C:/Users/runneradmin/go/pkg/mod/github.com/cockroachdb/datadriven@v1.0.3-0.20230413201302-be42291fc80f/datadriven.go:341
2024-02-12T20:39:05.8684012Z         	            				C:/Users/runneradmin/go/pkg/mod/github.com/cockroachdb/datadriven@v1.0.3-0.20230413201302-be42291fc80f/datadriven.go:210
2024-02-12T20:39:05.8685272Z         	            				C:/Users/runneradmin/go/pkg/mod/github.com/cockroachdb/datadriven@v1.0.3-0.20230413201302-be42291fc80f/datadriven.go:183
2024-02-12T20:39:05.8686523Z         	            				C:/Users/runneradmin/go/pkg/mod/github.com/cockroachdb/datadriven@v1.0.3-0.20230413201302-be42291fc80f/datadriven.go:143
2024-02-12T20:39:05.8687815Z         	            				C:/Users/runneradmin/go/pkg/mod/github.com/cockroachdb/datadriven@v1.0.3-0.20230413201302-be42291fc80f/datadriven.go:116
2024-02-12T20:39:05.8688346Z         	            				D:/a/pebble/pebble/wal/failover_writer_test.go:175
2024-02-12T20:39:05.8689627Z         	            				C:/Users/runneradmin/go/pkg/mod/github.com/cockroachdb/datadriven@v1.0.3-0.20230413201302-be42291fc80f/datadriven.go:424
2024-02-12T20:39:05.8690928Z         	            				C:/Users/runneradmin/go/pkg/mod/github.com/cockroachdb/datadriven@v1.0.3-0.20230413201302-be42291fc80f/datadriven.go:435
2024-02-12T20:39:05.8692188Z         	            				C:/Users/runneradmin/go/pkg/mod/github.com/cockroachdb/datadriven@v1.0.3-0.20230413201302-be42291fc80f/datadriven.go:448
2024-02-12T20:39:05.8693495Z         	            				C:/Users/runneradmin/go/pkg/mod/github.com/cockroachdb/datadriven@v1.0.3-0.20230413201302-be42291fc80f/test_interface.go:27
2024-02-12T20:39:05.8693708Z         	Error:      	"0s" is not less than "0s"
2024-02-12T20:39:05.8693957Z         	Test:       	TestFailoverWriter/errors
2024-02-12T20:39:05.8694115Z --- FAIL: TestFailoverWriter (0.19s)
2024-02-12T20:39:05.8694351Z     --- FAIL: TestFailoverWriter/blocking (0.19s)
2024-02-12T20:39:05.8694568Z     --- FAIL: TestFailoverWriter/errors (0.00s)
```

--- a/compaction_test.go
+++ b/compaction_test.go
@@ -14,7 +14,6 @@ import (
 	"path/filepath"
 	"reflect"
 	"regexp"
-	"runtime"
 	"slices"
 	"sort"
 	"strconv"
@@ -30,6 +29,7 @@ import (
 	"github.com/cockroachdb/pebble/internal/keyspan"
 	"github.com/cockroachdb/pebble/internal/manifest"
 	"github.com/cockroachdb/pebble/internal/testkeys"
+	"github.com/cockroachdb/pebble/internal/testutils"
 	"github.com/cockroachdb/pebble/objstorage"
 	"github.com/cockroachdb/pebble/objstorage/objstorageprovider"
 	"github.com/cockroachdb/pebble/objstorage/remote"
@@ -3048,14 +3048,9 @@ func TestCompactFlushQueuedMemTableAndFlushMetrics(t *testing.T) {
 				// The writes (during which the flush is idle) and the flush work
 				// should not be so fast as to be unrealistic. If these turn out to be
 				// flaky we could instead inject a clock.
-				//
-				// Windows timer precision is bad (on the order of 1 millisecond) and
-				// can cause the duration to be 0.
-				if runtime.GOOS != "windows" {
-					tinyInterval := 50 * time.Microsecond
-					require.Less(t, tinyInterval, metrics.Flush.WriteThroughput.WorkDuration)
-					require.Less(t, tinyInterval, metrics.Flush.WriteThroughput.IdleDuration)
-				}
+				tinyInterval := 50 * time.Microsecond
+				testutils.DurationIsAtLeast(t, metrics.Flush.WriteThroughput.WorkDuration, tinyInterval)
+				testutils.DurationIsAtLeast(t, metrics.Flush.WriteThroughput.IdleDuration, tinyInterval)
 				break
 			}
 			if time.Since(begin) > 2*time.Second {

--- a/internal/testutils/duration.go
+++ b/internal/testutils/duration.go
@@ -1,0 +1,25 @@
+// Copyright 2024 The LevelDB-Go and Pebble Authors. All rights reserved. Use
+// of this source code is governed by a BSD-style license that can be found in
+// the LICENSE file.
+
+package testutils
+
+import (
+	"runtime"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+)
+
+// DurationIsAtLeast verifies that the given duration is at least the given
+// value.
+func DurationIsAtLeast(t testing.TB, d, minValue time.Duration) {
+	t.Helper()
+	if runtime.GOOS == "windows" && minValue < 10*time.Millisecond {
+		// Windows timer precision is coarse (on the order of 1 millisecond) and can
+		// cause the duration for short operations to be 0.
+		return
+	}
+	require.GreaterOrEqual(t, d, minValue)
+}

--- a/wal/failover_writer_test.go
+++ b/wal/failover_writer_test.go
@@ -1,3 +1,7 @@
+// Copyright 2024 The LevelDB-Go and Pebble Authors. All rights reserved. Use
+// of this source code is governed by a BSD-style license that can be found in
+// the LICENSE file.
+
 package wal
 
 import (
@@ -6,7 +10,6 @@ import (
 	"io"
 	"math"
 	"math/rand"
-	"runtime"
 	"slices"
 	"strings"
 	"sync"
@@ -16,6 +19,7 @@ import (
 	"github.com/cockroachdb/datadriven"
 	"github.com/cockroachdb/errors"
 	"github.com/cockroachdb/pebble/internal/base"
+	"github.com/cockroachdb/pebble/internal/testutils"
 	"github.com/cockroachdb/pebble/record"
 	"github.com/cockroachdb/pebble/vfs"
 	"github.com/cockroachdb/pebble/vfs/errorfs"
@@ -32,9 +36,6 @@ const (
 )
 
 func TestFailoverWriter(t *testing.T) {
-	if runtime.GOOS == "windows" {
-		t.Skipf("flaky on windows (#3289)")
-	}
 	datadriven.Walk(t, "testdata/failover_writer", func(t *testing.T, path string) {
 		memFS := vfs.NewStrictMem()
 		dirs := [numDirIndices]dirAndFileHandle{
@@ -228,7 +229,7 @@ func TestFailoverWriter(t *testing.T) {
 						metrics := w.Metrics()
 						fmt.Fprintf(&b, "write bytes metric: %d\n", metrics.WriteThroughput.Bytes)
 						if metrics.WriteThroughput.Bytes > 0 {
-							require.Less(t, time.Duration(0), metrics.WriteThroughput.WorkDuration)
+							testutils.DurationIsAtLeast(t, metrics.WriteThroughput.WorkDuration, time.Nanosecond)
 						}
 					}
 					if stopGoroutines {

